### ISSUE TITLE
Enabling cuda GPU acceleration

### ISSUE THREAD

**Issue by klosax** [leejet/stable-diffusion.cpp#6]
It is possible to use cuBLAS by enabling it when compiling:
`-DGGML_CUBLAS=ON`

Maybe add this to the readme?

**Comment by Green-Sky** [comment]
did you run it/benchmark it?

**Comment by klosax** [comment]
On my system with 6 core Ryzen 5 and GTX 1660

w/o blas 30.2s
openblas 29.3s
cublas 16.6s

**Comment by klosax** [comment]
Setting openblas threads using the environment variable `export OPENBLAS_NUM_THREADS=6`
gets the number down to 19.86s

**Comment by klosax** [comment]
decode_first_stage timings

w/o blas, --threads=6 : 57.29s
openblas (threads=6), --threads 6 : 55.90s
cublas, --threads 6 : 52.57s

**Comment by klosax** [comment]
It looks like openblas gives some spped boost but only if setting the environment variable.
Enabling cublas will give a little more speed boost

**Comment by klosax** [comment]
It may be possible to find an optimal setting by testing different combinations of openblas threads and --threads. I guess this also applies to llama.cpp when using openblas.

**Comment by leejet** [comment]
> It is possible to use cuBLAS by enabling it when compiling: `-DGGML_CUBLAS=ON`
> 
> Maybe add this to the readme?

GPU support is already in my TODO list, and I'm working on adding it. However, I need to make ggml_conv_2d works on the GPU first.

**Comment by leejet** [comment]
> Time for decode_first_stage stays at about 56s in all tests.

This is expected as ggml conv 2d will not be optimized by blas and will not run on GPU, I am working on this issue

**Comment by leejet** [comment]
> It may be possible to find an optimal setting by testing different combinations of openblas threads and --threads. I guess this also applies to llama.cpp when using openblas.

Because ggml's thread is always busy waiting, even if no computation task is performed. As a result, it competes with blas threads, sometimes resulting in negative optimization. This is the point that needs to be optimized.

**Comment by ggerganov** [comment]
> This is expected as ggml conv 2d will not be optimized by blas and will not run on GPU, I am working on this issue

Btw, I'm not sure if the CPU version of conv 2d is optimal - most likely it is not.
There might be additional improvements possible if implemented properly

**Comment by klosax** [comment]
> GPU support is already in my TODO list, and I'm working on adding it. However, I need to make ggml_conv_2d works on the GPU first.

Great. But using cuBLAS is currently better than anything else if you have a cuda GPU.

> Because ggml's thread is always busy waiting, even if no computation task is performed. As a result, it competes with blas threads, sometimes resulting in negative optimization. This is the point that needs to be optimized.

Dont know if OpenBlas will have any real benefits over building without blas. And it looks like support for OpenBlas will soon be removed. See https://github.com/ggerganov/llama.cpp/pull/2372

**Comment by Happenedtostumblein** [comment]
@klosax If not, I suspect something is off with your benchmarks because it seems like the speed gain should be much higher with GPU vs CPU. Am I wrong?

Using -DGGML_OPENBLAS and applying the patch provided by ggml creator in #48, the GPU does get activated BUT CPU temps don’t drop by much. GPU stays hot after completion, so maybe those flags are not really doing anything.

Did yesterday’s cuda update affect your speed at all?

**Comment by LeonNerd** [comment]
> 

so,i have a question, the project  can only run on the cpu?i found vram is low when use cuBLAS by enabling it:
-DGGML_CUBLAS=ON and It has been a long time when use FP16 model.
what should i do for better infer time.

**Comment by juniofaathir** [comment]
@LeonNerd I think the main reason of this project exist is can run with CPU only + low RAM. But when we look at README.md, GPU inference is still on development.

And for shorter generation time, well maybe just generate 512 x 512 image only(don't use any BLAS), or get better CPU(?)

**Comment by LeonNerd** [comment]
> 
Yes, I looked at the TODO list. However, my test result 150s/iteration is  different from the above test result. It takes up 1g vram, 4g memory, 16 cors, and the gpu is A100.

**Comment by LeonNerd** [comment]
> @LeonNerd I think the main reason of this project exist is can run with CPU only + low RAM. But when we look at README.md, GPU inference is still on development.
> 
> And for shorter generation time, well maybe just generate 512 x 512 image only(don't use any BLAS), or get better CPU(?)
sure,i generated 512 x 512 image.I've found that the amount of time spent is related to the number of threads, not the more the better

**Comment by LeonNerd** [comment]
Can we get some inspiration from clip.cpp?

**Comment by FSSRepo** [comment]
@LeonNerd You can now activate the CUDA backend with `-DSD_CUBLAS=ON` and close this issue.

**Comment by Green-Sky** [comment]
@klosax would be cool if you could rerun the benchmarks :wink:

**Comment by leejet** [comment]
Support for CUDA GPU acceleration has already been added.
